File: realtime-macos/Sources/PAPAluLive/MouthGate.swift
enum MouthState: Equatable {
    case idle
    case talking
}

struct MouthGateConfiguration: Equatable {
    static let `default` = MouthGateConfiguration(
        openThreshold: 0.012,
        closeThreshold: 0.006,
        smoothingFactor: 0.35,
        releaseDelay: 0.60
    )

    let openThreshold: Double
    let closeThreshold: Double
    let smoothingFactor: Double
    let releaseDelay: Double

    init(
        openThreshold: Double,
        closeThreshold: Double,
        smoothingFactor: Double,
        releaseDelay: Double
    ) {
        precondition(openThreshold > closeThreshold)
        precondition(closeThreshold >= 0)
        precondition((0...1).contains(smoothingFactor))
        precondition(releaseDelay >= 0)

        self.openThreshold = openThreshold
        self.closeThreshold = closeThreshold
        self.smoothingFactor = smoothingFactor
        self.releaseDelay = releaseDelay
    }
}

struct MouthGate {
    private let configuration: MouthGateConfiguration
    private var smoothedRms = 0.0
    private var hasSample = false
    private var quietDuration = 0.0
    private(set) var state = MouthState.idle

    init(configuration: MouthGateConfiguration = .default) {
        self.configuration = configuration
    }

    mutating func update(rms: Double, duration: Double) -> MouthState {
        let sample = rms.isFinite ? max(0, rms) : 0
        let elapsed = duration.isFinite ? max(0, duration) : 0

        if hasSample {
            let alpha = configuration.smoothingFactor
            smoothedRms = alpha * sample + (1 - alpha) * smoothedRms
        } else {
            smoothedRms = sample
            hasSample = true
        }

        switch state {
        case .idle:
            if sample >= configuration.openThreshold {
                state = .talking
                quietDuration = 0
            }
        case .talking:
            if smoothedRms < configuration.closeThreshold {
                quietDuration += elapsed
                if quietDuration >= configuration.releaseDelay {
                    state = .idle
                    quietDuration = 0
                }
            } else {
                quietDuration = 0
            }
        }

        return state
    }
}

File: realtime-windows/PAPAluLive.Core/MouthGate.cs
namespace PAPAluLive.Core;

public enum MouthState
{
    Idle,
    Talking,
}

public sealed record MouthGateConfiguration(
    double OpenThreshold,
    double CloseThreshold,
    double SmoothingFactor,
    double ReleaseDelay)
{
    public static MouthGateConfiguration Default { get; } = new(
        OpenThreshold: 0.012,
        CloseThreshold: 0.010,
        SmoothingFactor: 0.35,
        ReleaseDelay: 0.60);

    public void Validate()
    {
        if (OpenThreshold <= CloseThreshold ||
            CloseThreshold < 0 ||
            SmoothingFactor is < 0 or > 1 ||
            ReleaseDelay < 0)
        {
            throw new ArgumentOutOfRangeException(
                nameof(MouthGateConfiguration));
        }
    }
}

public sealed class MouthGate
{
    private readonly MouthGateConfiguration configuration;
    private double smoothedRms;
    private bool hasSample;
    private double quietDuration;

    public MouthState State { get; private set; } = MouthState.Idle;

    public MouthGate(MouthGateConfiguration? configuration = null)
    {
        this.configuration = configuration ?? MouthGateConfiguration.Default;
        this.configuration.Validate();
    }

    public MouthState Update(double rms, double duration)
    {
        var sample = double.IsFinite(rms) ? Math.Max(0, rms) : 0;
        var elapsed = double.IsFinite(duration) ? Math.Max(0, duration) : 0;

        if (hasSample)
        {
            var alpha = configuration.SmoothingFactor;
            smoothedRms = alpha * sample + (1 - alpha) * smoothedRms;
        }
        else
        {
            smoothedRms = sample;
            hasSample = true;
        }

        switch (State)
        {
            case MouthState.Idle:
                if (sample >= configuration.OpenThreshold)
                {
                    State = MouthState.Talking;
                    quietDuration = 0;
                }
                break;
            case MouthState.Talking:
                if (smoothedRms < configuration.CloseThreshold)
                {
                    quietDuration += elapsed;
                    if (quietDuration >= configuration.ReleaseDelay)
                    {
                        State = MouthState.Idle;
                        quietDuration = 0;
                    }
                }
                else
                {
                    quietDuration = 0;
                }
                break;
            default:
                throw new InvalidOperationException("Unknown mouth state.");
        }

        return State;
    }
}
